sentry-android-replay/src/main/java/io/sentry/android/replay/ReplayCache.kt

```diff
@@ -0,0 +1,181 @@
+import io.sentry.android.replay.video.SimpleVideoEncoder
+import io.sentry.protocol.SentryId
+import java.io.Closeable
+import java.io.File
+
+public class ReplayCache internal constructor(
+    private val options: SentryOptions,
+    private val replayId: SentryId,
+    private val recorderConfig: ScreenshotRecorderConfig,
+    private val encoderCreator: (File) -> SimpleVideoEncoder
+) : Closeable {
+
+    public constructor(
+        options: SentryOptions,
+        replayId: SentryId,
+        recorderConfig: ScreenshotRecorderConfig
+    ) : this(options, replayId, recorderConfig, encoderCreator = { videoFile ->
+        SimpleVideoEncoder(
+            options,
+            MuxerConfig(
+                file = videoFile,
+                recorderConfig = recorderConfig,
+                frameRate = recorderConfig.frameRate.toFloat(),
+                bitrate = 20 * 1000
+            )
+        ).also { it.start() }
+    })
+
+    private val encoderLock = Any()
+    private var encoder: SimpleVideoEncoder? = null
+
+    internal val replayCacheDir: File? by lazy {
+        if (options.cacheDirPath.isNullOrEmpty()) {
+            options.logger.log(
+                WARNING,
+                "SentryOptions.cacheDirPath is not set, session replay is no-op"
+            )
+            null
+        } else {
+            File(options.cacheDirPath!!, "replay_$replayId").also { it.mkdirs() }
+        }
+    }
+
+    // TODO: maybe account for multi-threaded access
+    internal val frames = mutableListOf<ReplayFrame>()
+
+    internal fun addFrame(bitmap: Bitmap, frameTimestamp: Long) {
+        if (replayCacheDir == null) {
+            return
+        }
+
+        val screenshot = File(replayCacheDir, "$frameTimestamp.jpg").also {
+            it.createNewFile()
+        }
+        screenshot.outputStream().use {
+            bitmap.compress(JPEG, 80, it)
+            it.flush()
+        }
+
+        addFrame(screenshot, frameTimestamp)
+    }
+
+    public fun addFrame(screenshot: File, frameTimestamp: Long) {
+        val frame = ReplayFrame(screenshot, frameTimestamp)
+        frames += frame
+    }
+
+    public fun createVideoOf(
+        duration: Long,
+        from: Long,
+        segmentId: Int,
+        videoFile: File = File(replayCacheDir, "$segmentId.mp4")
+    ): GeneratedVideo? {
+        if (frames.isEmpty()) {
+            options.logger.log(
+                DEBUG,
+                "No captured frames, skipping generating a video segment"
+            )
+            return null
+        }
+
+        // TODO: reuse instance of encoder and just change file path to create a different muxer
+        encoder = synchronized(encoderLock) { encoderCreator(videoFile) }
+
+        val step = 1000 / recorderConfig.frameRate.toLong()
+        var frameCount = 0
+        var lastFrame: ReplayFrame = frames.first()
+        for (timestamp in from until (from + (duration)) step step) {
+            val iter = frames.iterator()
```

Comment:
this could definitely cause `ConcurrentModificationException` issues in case a frame is added in the meantime

Reply:
for sure, I kind of didn't care about that because I use a single-thread executor for calling these methods.

Do you think it'd be enough to just leave comments and say this class is not thread-safe (because there's no way to enforce it really), or we still should make it thread-safe?